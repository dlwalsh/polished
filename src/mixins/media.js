function pxNum(value: string | number): string {
  return typeof value === 'number' ? `${value}px` : value
}

/**
 * Returns a media query according to provided arguments
 *
 * @example
 * // styled-components usage
 * const div = styled.div`
 *   width: 100%;
 *   ${media({ minWidth: "900px" })} {
 *     width: 50%;
 *   }
 *   ${media({ minWidth: "1200px" })} {
 *      width: 33.33%;
 *   }
 * `
 *
 * // CSS as JS Output
 *
 * div: {
 *   "width": "100%",
 *   "@media (min-width: 900px)": Object {
 *     "width": "50%",
 *   },
 *   "@media (min-width: 1200px)": Object {
 *     "width": "33.33%",
 *   },
 * }
 */

function media(query: string | Object): string {
  if (typeof query === 'string') {
    return `@media ${query}`
  }

  if (typeof query !== 'object') {
    throw new Error('Property must be an object or string')
  }

  const { type, minWidth, maxWidth } = query
  const minWidthStr = minWidth ? `(min-width: ${pxNum(minWidth)})` : undefined
  const maxWidthStr = maxWidth ? `(max-width: ${pxNum(maxWidth)})` : undefined

  const queryStr = [type, minWidthStr, maxWidthStr].filter(x => x).join(' and ')

  return `@media ${queryStr}`
}

export default media
