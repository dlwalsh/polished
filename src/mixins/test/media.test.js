import media from '../media'

describe('media', () => {
  it('should return a valid media query when passed min width', () => {
    expect(
      media({
        minWidth: 100,
      }),
    ).toMatchSnapshot()
  })

  it('should return a valid media query when passed max width', () => {
    expect(
      media({
        maxWidth: 600,
      }),
    ).toMatchSnapshot()
  })

  it('should return a valid media query when passed min/max widths', () => {
    expect(
      media({
        minWidth: 100,
        maxWidth: 600,
      }),
    ).toMatchSnapshot()
  })

  it('should return a valid media query when passed type', () => {
    expect(
      media({
        type: 'print',
      }),
    ).toMatchSnapshot()
  })

  it('should return a valid media query when passed type and min/max widths', () => {
    expect(
      media({
        type: 'all',
        minWidth: '10rem',
        maxWidth: '30rem',
      }),
    ).toMatchSnapshot()
  })

  it('should return a valid media query when passed a string', () => {
    expect(media('(orientation: portrait)')).toMatchSnapshot()
  })

  it('should throw an error when passed a boolean', () => {
    expect(() => {
      media(true)
    }).toThrow('Property must be an object or string')
  })
})
