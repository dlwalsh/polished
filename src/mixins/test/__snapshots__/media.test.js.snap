// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media should return a valid media query when passed a string 1`] = `"@media (orientation: portrait)"`;

exports[`media should return a valid media query when passed max width 1`] = `"@media (max-width: 600px)"`;

exports[`media should return a valid media query when passed min width 1`] = `"@media (min-width: 100px)"`;

exports[`media should return a valid media query when passed min/max widths 1`] = `"@media (min-width: 100px) and (max-width: 600px)"`;

exports[`media should return a valid media query when passed type 1`] = `"@media print"`;

exports[`media should return a valid media query when passed type and min/max widths 1`] = `"@media all and (min-width: 10rem) and (max-width: 30rem)"`;
